src/basic/io-util.c

```diff
@@ -159,21 +158,29 @@ int pipe_eof(int fd) {
         return !!(r & POLLHUP);
 }
 
-int fd_wait_for_event(int fd, int event, usec_t t) {
+int ppoll_usec(struct pollfd *fds, size_t nfds, usec_t timeout) {
+        struct timespec ts;
+        int r;
+
+        assert(fds);
+
+        r = ppoll(fds, nfds, timeout == USEC_INFINITY ? NULL : timespec_store(&ts, timeout), NULL);
+        if (r < 0)
+                return -errno;
 
+        return r;
+}
```

Comment:
this looks like a great simplification. but i'd go one step further and check for POLLNVAL here in this function too. Pretty much any user of this is going to do that (or should at least), hence we should just do this implicitly. i.e. just iterate through the array after poll() returns and convetr POLLNVAL into -EBADF. To optimize this a bit it might make sense to abort the loop once enough entries with revents != 0 have been seen matching what ppoll() returned.

Reply:
> To optimize this a bit it might make sense to abort the loop once enough entries with revents != 0 have been seen matching what ppoll() returned.

Hmm? Sorry, I cannot get this. Could you elaborate more?

Reply:
ppoll() returns how many entries in the fds[] array have .revents != 0. Under the assumption that pretty often this will be a few only (ie. often just one) it might make sense to exit the loop early once we seen the right number of them. thus cutting the number of iterations to half in average. i.e.:

```c
r = ppoll(…);
if (r < 0)
        …
for (size_t i = 0; i < nfds; i++) {
        if (r == 0)
                break;

        if (fds[i].revents & POLLNVAL)
                return -EBADF;

        r--;
}
```

Reply:
Thanks.